File: app/src/main/java/com/cbr/behancesampleapp/di/AppComponent.java
package com.cbr.behancesampleapp.di;

import android.app.Application;

import com.cbr.behancesampleapp.BehanceSampleApp;
import com.cbr.behancesampleapp.di.module.ActivityGraphBinderModule;
import com.cbr.behancesampleapp.di.module.AppModule;
import com.cbr.behancesampleapp.di.module.NetModule;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjectionModule;
import dagger.android.AndroidInjector;
import dagger.android.DaggerApplication;

/**
 * Created by dimitrios on 22/08/2017.
 * <p>
 * Taken From https://github.com/googlesamples/android-architecture/blob/todo-mvp-dagger/
 * <p>
 * This is a Dagger component. Refer to {@link com.cbr.behancesampleapp.BehanceSampleApp} for the list of Dagger components
 * used in this application.
 * <p>
 * Even though Dagger allows annotating a {@link Component} as a singleton, the code
 * itself must ensure only one instance of the class is created. This is done in {@link com.cbr.behancesampleapp.BehanceSampleApp}.
 * <p>
 * //{@link AndroidInjectionModule}
 * // is the module from Dagger.Android that helps with the generation
 * // and location of subcomponents.
 */
@Singleton
@Component(modules = {
	AndroidInjectionModule.class,
	ActivityGraphBinderModule.class,
	AppModule.class,
	NetModule.class})
public interface AppComponent extends AndroidInjector<DaggerApplication> {

	@Component.Builder
	interface Builder {

		@BindsInstance
		Builder application(Application application);

		AppComponent build();
	}

	void inject(BehanceSampleApp application);

}

File: app/src/main/java/com/cbr/behancesampleapp/dagger/AppComponent.kt
package com.cbr.behancesampleapp.dagger

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.content.res.Resources
import com.cbr.behancesampleapp.dagger.module.AppModule
import com.cbr.behancesampleapp.dagger.module.NetModule
import com.cbr.behancesampleapp.domain.network.BehanceApiService
import com.cbr.behancesampleapp.domain.network.BehanceRepository
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.DaggerApplication
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import javax.inject.Singleton

@Singleton
@Component(modules = [(AppModule::class), (NetModule::class)])
interface AppComponent {
    
    fun context(): Context
    
    fun resources(): Resources
    
    fun sharedPrefs(): SharedPreferences
    
    fun cache(): Cache
    
    fun okHttpInterceptor(): Interceptor
    
    fun okHttpClient(): OkHttpClient
    
    fun retrofit(): Retrofit
    
    fun behanceService(): BehanceApiService
    
    fun behanceRepository(): BehanceRepository
}
